File: spec/Airport_Spec.rb
require 'Airport.rb'


describe Airport do

it { is_expected.to respond_to(:land).with(1).argument }
it { is_expected.to respond_to :planes_landed }
it { is_expected.to respond_to(:takeoff).with(1).argument}
it { is_expected.to respond_to(:current_weather)}
it { is_expected.to respond_to(:capacity)}
it { is_expected.to respond_to(:capacity_hit)}

let(:plane){double :plane}
let(:plane2){double :plane2}

it 'has a default capacity' do
  expect(subject.capacity).to eq Airport::DEFAULT_CAPACITY
  end

it 'lands a plane' do
  plane = double(:plane)
  airport = Airport.new
  allow(airport).to receive(:current_weather).and_return(false)
  expect(airport.land(plane)).to eq [plane]
end

it 'cannot land a plane more than once' do
  plane = double(:plane)
  airport = Airport.new
  allow(airport).to receive(:current_weather).and_return(false)
  airport.land(plane)
  expect {airport.land(plane)}.to raise_error ("this plane has already landed")
end


it "shows plane has landed" do
  plane = double(:plane)
  airport = Airport.new
  allow(airport).to receive(:current_weather).and_return(false)
  airport.land(plane)
  expect(airport.planes_landed).to eq [(plane)]
  end

it 'lets a specified plane takeoff' do
  plane = double(:plane)
  plane2 = double(:plane2)
  airport = Airport.new
  allow(airport).to receive(:current_weather).and_return(false)
  airport.land(plane)
  airport.land(plane2)
  expect(airport.takeoff(plane)).to eq [plane]
end


it 'makes sure you cannot tell a plane to takeoff more than once' do
  plane = double(:plane)
  airport = Airport.new
  allow(airport).to receive(:current_weather).and_return(false)
  airport.land(plane)
  airport.takeoff(plane)
  expect {airport.takeoff(plane)}.to raise_error ("this plane has already taken off")
end

it 'prevents takeoff when weather is stormy' do
  plane = double(:plane)
  airport = Airport.new
  allow(airport).to receive(:current_weather).and_return(false)
  airport.land(plane)
  allow(airport).to receive(:current_weather).and_return(true)
  expect {airport.takeoff(plane)}.to raise_error ("plane cannot take off due to stormy weather")
end

it 'prevents landing when weather is stormy' do
   plane = double(:plane)
   airport = Airport.new
   allow(airport).to receive(:current_weather).and_return(true)
   expect {airport.land(plane)}.to raise_error ("plane cannot land due to stormy conditions")
 end

it 'prevents landing if the airport is at capacity' do
   plane = double(:plane)
   airport = Airport.new
   allow(airport).to receive(:capacity_hit).and_return(true)
   allow(airport).to receive(:current_weather).and_return(false)
   expect {airport.land(plane)}.to raise_error ("plane cannot land as airport is full")
end



end

File: spec/airportSpec.js
describe('airport', function() {
  let airport;
  let plane;

  beforeEach(function() {
    airport = new Airport();
    plane = new Plane();
  });

  describe('runway', function() {
    it('has no planes by default', function() {
      expect(airport.runway).toEqual([]);
    });
  });

  describe('land', function() {
    it('can land at an airport', function() {
      expect(airport.land).not.toBeUndefined();
    });

    it('a plane can land', function() {
      spyOn(airport, 'isStormy').and.returnValue(false);
      airport.land(plane)
      expect(airport.runway).toEqual([plane]);
    });

    it('throws error if airport has reached capacity', function() {
      spyOn(airport, 'isStormy').and.returnValue(false);
      airport.capacity = 1;
      airport.land(plane);
      expect( function() { airport.land(plane); } ).toThrow(new Error('Airport Full'));
    });
  });

  describe('takeoff', function() {
    it('can takeoff from the airport', function() {
      expect(airport.takeoff).not.toBeUndefined();
    });

    it('throws error if no planes to takeoff', function() {
      spyOn(airport, 'isStormy').and.returnValue(false);
      expect(airport.runway).toEqual([])
      expect( function() { airport.takeoff(); } ).toThrow(new Error('Airport Empty'));
    });

    it('a plane can takeoff', function() {
      spyOn(airport, 'isStormy').and.returnValue(false);
      airport.land(plane);
      airport.takeoff();
      expect(airport.runway).not.toContain(plane);
    });
  });

  describe('capacity', function() {
    it('has a capcity', function() {
      expect(airport.capacity).not.toBeUndefined();
    });

    it('has a default capacity of 20', function() {
      expect(airport.capacity).toEqual(20);
    });

    it('has a variable capacity', function() {
      airport.capacity = 30;
      expect(airport.capacity).toEqual(30);
    });
  });

  describe('weather', function() {
    it('has an isStormy function', function() {
      expect(airport.isStormy()).not.toBeUndefined();
    });

    it('stops takeoff when weather is stormy', function() {
      airport.land(plane);
      spyOn(airport, 'isStormy').and.returnValue(true);
      expect( function() { airport.takeoff(); } ).toThrow(new Error('Cannot Takeoff In This Weather'));
      expect(airport.runway).toContain(plane);
    });

    it('allows takeoff when weather is not stormy', function() {
      spyOn(airport, 'isStormy').and.returnValue(false);
      airport.land(plane);
      airport.takeoff();
      expect(airport.runway).not.toContain(plane);
    });
  });

  it('stops landing when weather is stormy', function() {
    spyOn(airport, 'isStormy').and.returnValue(true);
    expect( function() { airport.land(plane); } ).toThrow(new Error('Cannot Land In This Weather'));
  });

  it('allows landing when weather is not stormy', function() {
    spyOn(airport, 'isStormy').and.returnValue(false);
    airport.land(plane);
    expect(airport.runway).toEqual([plane]);
  });
});